### PR TITLE
✅ Add unit test to find missing translation in english

### PR DESCRIPTION
**Pull request by coliss86**
ping @PPAC37, this PR is heavily inspired from your https://github.com/MarginallyClever/Makelangelo-software/pull/456

Here what happened when a translation is missing: 
```
21:48:01.532 [main] INFO  com.marginallyclever.makelangelo.MissingTranslationTest - translations missing:
21:48:01.534 [main] INFO  com.marginallyclever.makelangelo.MissingTranslationTest -   src\main\java\com\marginallyclever\makelangelo\makeArt\imageConverter\Converter_CannyEdge.java:20: CannyEdgeConverterName
21:48:01.535 [main] INFO  com.marginallyclever.makelangelo.MissingTranslationTest -   src\main\java\com\marginallyclever\makelangelo\makeArt\imageConverter\Converter_ColorFloodFill.java:45: RGBFloodFillName
...
```

But first, I have to do another PR the fix missing translation :)

### CODE REVIEW

**Comment by PPAC37** [comment]
Really better as it is simpler than my https://github.com/MarginallyClever/Makelangelo-software/pull/456

as you are basing yourself on my code, this also returns the Traduction.get in comment
https://regex101.com/r/g4V0uw/1

(simple to modify the regexp (thanks to the one found in https://stackoverflow.com/questions/25425755/regular-expression-to-find-non-commented-java-lines see a test applicable to my current implementation with some simple twick (but necessarily which like line by line only half works...)
https://regex101.com/r/1kHY21/1
so as not to retain those which are in a "//" but not for those in a "/* */" on several lines with the implementation I made (reading line by line).
(There's also https://www.debuggex.com/ as an online tool for reg exp (I like state-rendered to understand what a regexp is doing...))

    Translator\s*\.\s*get\s*\(([^\\)]*)\)|(['"])(?:(?!\1|\\).|\\. )*\1|\/\/[^\n]*(?:\n|$)|\/\*(?:[^*]|\*(?!\/))*\*\/

Regular expression visualization

[Debuggex Demo](https://www.debuggex.com/r/9MioNPd3NTsa4t10)


Otherwise I had looked at the possibility of generating an issue ( https://github.com/marketplace/actions/create-issue-from-file created or completed with the content of a file if it exists ), I did tests on my fork with one of my branches https://github.com/PPAC37/Makelangelo-software/tree/searchInTheCode TraductionGetToFindTheKey_ci_create_issues but I haven't yet studied the trick not to create an issue, each time a build is done...

**Comment by i-make-robots** [comment]
added missing strings to master.

**Comment by coliss86** [comment]
> Really better as it is simpler than my #456
> 
> as you are basing yourself on my code, this also returns the Traduction.get in comment https://regex101.com/r/g4V0uw/1
> 
> (simple to modify the regexp (thanks to the one found in https://stackoverflow.com/questions/25425755/regular-expression-to-find-non-commented-java-lines see a test applicable to my current implementation with some simple twick (but necessarily which like line by line only half works...) https://regex101.com/r/1kHY21/1 so as not to retain those which are in a "//" but not for those in a "/* */" on several lines with the implementation I made (reading line by line). (There's also https://www.debuggex.com/ as an online tool for reg exp (I like state-rendered to understand what a regexp is doing... )) https://www.debuggex.com/r/xzJMkqryBQgBWc4h

Ok, I'll implement it asap

> Otherwise I had looked at the possibility of generating an issue ( https://github.com/marketplace/actions/create-issue-from-file created or completed with the content of a file if it exists ), I did tests on my fork with one of my branches https://github.com/PPAC37/Makelangelo-software/tree/searchInTheCodeTraductionGetToFindTheKey_ci_create_issues but I haven't yet studied the trick not to create an issue, each time a build is done...

I don't think it's necessary as this test will be executed of each PR and will fails if a key is missing, so you won't be able to merge it and thus you'll have to fix the code before.

**Comment by coliss86** [comment]
@PPAC37 I have added a filter on lines starting with a comment.
There is no multi line comment for now in the code and I think it will be far more difficult to implement.

**Comment by PPAC37** [comment]
@coliss86  Weel done ! 
And yes, it's not easy to find the right algorithm with the regexp that meets our needs. (I broke my teeth on it, especially since I'm quickly unable to express what I want in regexp (ex: when I want a negation like do not containe "//") and the only method I've found is to add a | "or part" to the regexp and to make a special case in the treatment of the match (if the group which interests me is null then it is the match on the " | or part" and therefore nothing to do for this match ...) )

File: src/test/java/com/marginallyclever/makelangelo/MissingTranslationTest.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    @Test
+    public void verifyThatMatcherWorks_results() throws IOException {
+        List<String> results = new ArrayList<>();
+
+        // matches.txt contains few entries with translation and one without. The list should not be empty
+        searchInAFile(new File("src/test/resources/translator/matches.txt"), results);
+
+        assertEquals(List.of("src/test/resources/translator/matches.txt:4: unknownKey"), results);
```
Comment (PPAC37) [inline src/test/java/com/marginallyclever/makelangelo/MissingTranslationTest.java:68]: If you run it on a windows systeme ( file char separator "\\" vs "/" ) ... (this is one of the current CI test failed for this PR : 
~~~
Error:    MissingTranslationTest.verifyThatMatcherWorks_results:68 expected: <[src/test/resources/translator/matches.txt:4: unknownKey]> but was: <[src\test\resources\translator\matches.txt:4: unknownKey]>
~~~
)
maybe juste 
`"src/test/resources/translator/matches.txt:4: unknownKey".replaceAll("/", File.separator)` 
?
or else convert to URI the File for the output on the file ?

Comment (coliss86) [reply src/test/java/com/marginallyclever/makelangelo/MissingTranslationTest.java]: thx, I modified with `.replaceAll("/", File.separator)`

File: src/test/java/com/marginallyclever/makelangelo/MissingTranslationTest.java [hunk 2]
```diff
@@ -65,7 +65,7 @@ public void verifyThatMatcherWorks_results() throws IOException {
         // matches.txt contains few entries with translation and one without. The list should not be empty
         searchInAFile(new File("src/test/resources/translator/matches.txt"), results);
 
-        assertEquals(List.of("src/test/resources/translator/matches.txt:4: unknownKey"), results);
+        assertEquals(List.of("src/test/resources/translator/matches.txt:4: unknownKey".replaceAll("/", File.separator)), results);
```
Comment (PPAC37) [inline src/test/java/com/marginallyclever/makelangelo/MissingTranslationTest.java:68]: Sorry i may have miss something
now we have
~~~
Error:  Errors: 
Error:    MissingTranslationTest.verifyThatMatcherWorks_results:68 � IllegalArgument cha...
~~~
and I'm confused as to why. encoding problem of my message if you copy pasted it?

Comment (coliss86) [reply src/test/java/com/marginallyclever/makelangelo/MissingTranslationTest.java]: I changed the output of the test, on IntelliJ, I can now click in the log on the file and it opens directty to the file at the line. Also update the assertion to not fails on windows (hopefully)

image

**Pull request merged**